File: 2021/day03/part01/main.zig
const std = @import("std");
// const input = @embedFile("example.txt");
const input = @embedFile("input.txt");

pub fn to_decimal(bits: [12]usize) usize {
    var pow: usize = 11;
    var decimal: usize = 0;
    for (bits) |b| {
        decimal += b * std.math.pow(usize, 2, pow);
        if (pow >= 1) { pow -= 1; }
    }

    return decimal;
}

pub fn main() !void {
    var lines = std.mem.tokenize(u8, input, "\n");

    // create array of counts
    var counts = [1]isize{0} ** 12;
    // var counts = [1]isize{0} ** 5;
    while (lines.next()) |bits| {
        for (bits) |bit, index| {
            counts[index] += switch(bit) {
                '1' => @as(isize, 1),
                '0' => @as(isize, -1),
                else => unreachable,
            };
        }
    }

    // create arrays that represent gamma and epsilon binary numbers
    var epsilon_bits = [1]usize{0} ** 12;
    var gamma_bits = [1]usize{0} ** 12;
    for (counts) |count, index| {
        if (count > 0) { gamma_bits[index] = 1; }
        else { epsilon_bits[index] = 1; }
    }

    // convert to decimal numbers
    var gamma = to_decimal(gamma_bits);
    var epsilon = to_decimal(epsilon_bits); 
    std.debug.print("result: {}\n", .{epsilon * gamma});
}

File: 2021/day03/part01/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
)

var numLength = 12

func readFile(path string) []string {
	file, err := os.Open(path)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}

	return lines
}

func getPositionCounts(lines []string) []int {
	counts := make([]int, numLength)

	for _, line := range lines {
		for i, c := range line {
			switch c {
			case '1':
				counts[i] += 1
			case '0':
				counts[i] -= 1
			}
		}
	}

	return counts
}

func createBinaries(counts []int) ([]int, []int) {
	gammaBits := make([]int, numLength)
	epsilonBits := make([]int, numLength)

	for i, count := range counts {
		if count > 0 {
			gammaBits[i] = 1
		} else {
			epsilonBits[i] = 1
		}
	}

	return gammaBits, epsilonBits
}

func binaryArrToDecimal(binary []int) float64 {
	pow := float64(numLength - 1)
	decimal := 0.0
	for _, bit := range binary {
		var fBit = float64(bit)
		decimal += fBit * math.Pow(2, pow)
		if pow >= 1.0 {
			pow -= 1.0
		}
	}

	return decimal
}

func main() {
	// lines := readFile("example.txt")
	lines := readFile("input.txt")

	counts := getPositionCounts(lines)

	gammaBits, epsilonBits := createBinaries(counts)

	gamma := binaryArrToDecimal(gammaBits)
	epsilon := binaryArrToDecimal(epsilonBits)

	fmt.Printf("result: %d\n", int(gamma*epsilon))
}
